connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/WorkerErrantRecordReporter.java

```diff
@@ -0,0 +1,175 @@
+        Converter keyConverter,
+        Converter valueConverter,
+        HeaderConverter headerConverter
+    ) {
+        this.retryWithToleranceOperator = retryWithToleranceOperator;
+        this.keyConverter = keyConverter;
+        this.valueConverter = valueConverter;
+        this.headerConverter = headerConverter;
+        this.futures = new LinkedList<>();
+    }
+
+    @Override
+    public Future<Void> report(SinkRecord record, Throwable error) {
+        Function<SinkRecord, ConsumerRecord<byte[], byte[]>> function;
+
+        if (record instanceof InternalSinkRecord) {
+            function = sinkRecord -> ((InternalSinkRecord) sinkRecord).originalRecord();
+        } else {
+            function = sinkRecord -> {
+
+                String topic = record.topic();
+                byte[] key = keyConverter.fromConnectData(topic, record.keySchema(), record.key());
+                byte[] value = valueConverter.fromConnectData(topic, record.valueSchema(),
+                    record.value());
+
+                RecordHeaders headers = new RecordHeaders();
+                if (record.headers() != null) {
+                    for (Header header : record.headers()) {
+                        String headerKey = header.key();
+                        byte[] rawHeader = headerConverter.fromConnectHeader(topic, headerKey,
+                            header.schema(), header.value());
+                        headers.add(headerKey, rawHeader);
+                    }
+                }
+
+                return new ConsumerRecord<>(record.topic(), record.kafkaPartition(),
+                    record.kafkaOffset(), record.timestamp(), record.timestampType(), -1L, -1,
+                    -1, key, value, headers);
+
+            };
+        }
+
+        Future<Void> future = retryWithToleranceOperator.executeFailed(function, Stage.TASK_PUT,
+            SinkTask.class, record, error, callback);
+
+        futures.add(future);
+        return future;
+    }
+
+    /**
+     * Gets all futures returned by the sink records sent to Kafka by the errant
+     * record reporter. This function is intended to be used to block on all the errant record
+     * futures.
+     */
+    public void getAllFutures() {
+        for (Future<Void> future : futures) {
+            try {
+                future.get();
+            } catch (InterruptedException | ExecutionException e) {
+                log.error("Encountered an error while calling ");
+                throw new ConnectException(e);
+            }
+        }
+        futures.clear();
+    }
+
+    /**
+     * Wrapper class to aggregate producer futures and abstract away the record metadata from the
+     * Connect user.
+     */
+    public static class ErrantRecordFuture implements Future<Void> {
+
+        private final List<Future<RecordMetadata>> futures;
+
+
+        public ErrantRecordFuture(List<Future<RecordMetadata>> producerFutures) {
+            futures = producerFutures;
+        }
```

Comment:
Rather than have a list of futures, why not have a single `Future` delegate that is either a `CompletableFuture.allOf(...)` or a single feature? This makes the constructor a little more complex, but it would simplify all of the other methods tremendously since they merely have to delegate (except for `cancel()` and `isCancelled()`, which can stay the same:
```suggestion
        public ErrantRecordFuture(List<Future<RecordMetadata>> producerFutures) {
            if (producerFutures == null || producerFutures.isEmpty()) {
                future = CompletableFuture.completedFuture(null);
            } else {
                futures = CompletableFutures.allOf(producerFutures);
            }
        }
```
This will make `get(long, TimeUnit)` behave more correctly by requiring that all futures complete within the stated time.